bouter/free/__init__.py

```diff
@@ -1,6 +1,250 @@
+                "f{:d}_y".format(i_fish): "y",
+                "f{:d}_vy".format(i_fish): "vy",
+                "f{:d}_theta".format(i_fish): "theta",
+                "f{:d}_vtheta".format(i_fish): "vtheta",
+            },
+            **{
+                "f{:d}_theta_{:02d}".format(i_fish, i): "theta_{:02d}".format(i)
+                for i in range(n_segments)
+            }
+        )
+
+
+    def _rename_fish(self, df, i_fish, n_segments):
+        return df.filter(["t"] + self._fish_column_names(i_fish, n_segments)).rename(
+            columns=self._fish_renames(i_fish, n_segments)
+        )
+
+
+    @decorators.cache_results(cache_filename="behavior_log")
+    def compute_velocity(
+        self,
+        max_interpolate=2,
+        recalculate_vel=False,
+        scale=None
+    ):
+        """Compute the squared total swimming velocity for each fish.
+        Add them as new columns to the dataframe log and return the complete dataframe.
+
+        :param max_interpolate: number of points to interpolate if surrounded by NaNs in tracking
+        :param recalculate_vel:
+        :param scale: mm per pixel, recalculated by default
+        :return:
+        """
+        df = self.behavior_log
+        scale = scale or self.camera_px_in_mm
+        dt = np.mean(np.diff(df.t[100:200]))
+
+        n_fish = self.n_fish
+        dfint = df.interpolate("linear", limit=max_interpolate, limit_area="inside")
+        for i_fish in range(n_fish):
+            if recalculate_vel:
+                for thing in ["x", "y", "theta"]:
+                    dfint["f{}_v{}".format(i_fish, thing)] = np.r_[
+                        np.diff(dfint["f{}_{}".format(i_fish, thing)]), 0
+                    ]
+
+            vel2 = (
+                dfint["f{}_vx".format(i_fish)] ** 2 + dfint["f{}_vy".format(i_fish)] ** 2
+            ) * ((scale / dt) ** 2)
+
+            self.behavior_log["vel2_f{}".format(i_fish)] = vel2
+
+        return self.behavior_log
+
+
+    @decorators.cache_results()
+    def get_bouts(
+        self,
+        median_vel=False,
+        window_size=7,
+        scale=None,
+        threshold=1,
+        **kwargs
+    ):
+        """ Extracts all bouts from a freely-swimming tracking experiment
+
+        :param exp: the experiment object
+        :param scale: mm per pixel, recalculated by default
+        :param threshold: velocity threshold in mm/s
+        :return: tuple: (list of single bout dataframes, list of boolean arrays marking if the
+         bout i follows bout i-1)
+        """
+        n_fish = self.n_fish
+        n_segments = self.n_tail_segments
+        scale = scale or self.camera_px_in_mm
+
+        self.compute_velocity()
+
+        bouts = []
+        continuous = []
+
+        for i_fish in range(n_fish):
+            vel2 = self.behavior_log["vel2_f{}".format(i_fish)]
+            if median_vel:
+                vel2 = vel2.rolling(window=window_size, min_periods=1).median()
```

Comment:
this I would put in computing velocity, but it's a minor point. A more important point is that the absolute velocity of the fish is almost never needed, and should not be cached in my view (as it is also a quite cheap thing to recompute).